app/parse.py

```diff
@@ -15,8 +41,103 @@ class Product:
     num_of_reviews: int
 
 
+def parse_single_product(product_soup: Tag) -> Product:
+    return Product(
+        title=product_soup.select_one(".title")["title"],
+        description=product_soup.select_one(".description").text,
+        price=float(product_soup.select_one(".price").text.replace("$", "")),
+        rating=int(product_soup.select_one("p[data-rating]")["data-rating"]),
+        num_of_reviews=int(
+            product_soup.select_one("p.review-count").text.split()[0]
+        ),
+    )
+
+
+def save_to_csv(products: list[Product], file_name: str) -> None:
+    with open(file_name, "w", newline='') as f:
+        writer = csv.writer(f)
+        writer.writerow(FIELDS)
+        writer.writerows([astuple(product) for product in tqdm(products)])
+
+
+def parse_single_product_from_detail_page(url: str) -> Product:
+    tablet_url = urljoin(BASE_URL, url)
+    tablet_page = requests.get(tablet_url).content
+    tablet_soup = BeautifulSoup(tablet_page, "html.parser")
+
+    return Product(
+        title=tablet_soup.select_one(".title").text,
+        description=tablet_soup.select_one(".description").text,
+        price=float(tablet_soup.select_one(".price").text.replace("$", "")),
+        rating=len(tablet_soup.select("p.review-count span")),
```

Comment:
The calculation of the rating seems incorrect. It should be extracted from the 'data-rating' attribute similar to the 'parse_single_product' function.